### ISSUE TITLE
Generic interface inheriting from another generic interface results in incorrect inference

### ISSUE THREAD

**Issue by alshain** [Microsoft/TypeScript#18305]
**TypeScript Version:**  2.5/2.4.2/2.4.1/2.3

**Code**

```ts
interface Api<I, O> {
    url: string;
    method: 'POST' | 'GET';
}

interface GetApi<O> extends Api<{}, O> {
    method: 'GET';
}

type PostApi<I, O> = Api<I, O>;

function defineGetApi<O>(api: GetApi<O>) {
    return api;
}

function definePostApi<I, O>(api: PostApi<I, O>) {
    return api;
}


interface SampleInput {
    command: string;
}

interface SampleOutput {
    success: boolean;
}

const sampleGetApi = defineGetApi<SampleOutput>({
    method: 'GET',
    url: 'string',
});
const samplePostApi = definePostApi<SampleInput, SampleOutput>({
    method: 'GET',
    url: 'string',
});

function consumeApi<I, O>(api: Api<I, O>, input: I, onSuccess?: (output: O) => void) {
    return;
}

consumeApi(sampleGetApi, {}, (o) => o.success); //(1)
consumeApi(samplePostApi, {}, (o) => o); //(2)
```

**Expected behavior:**
(1) should typecheck
(2) should not type check, but I guess that's by design

**Actual behavior:**
(1) does not typecheck, the interface inheritance doesn't seem to be able to infer `O`.

> Property 'success' does not exist on type '{}'

With (2), using only a type alias, the inference for `O` works as expected. However, thanks to `{}` being a supertype of `SampleInput`... I still easily get bad code :(

![image](https://user-images.githubusercontent.com/349397/30156866-1cb24490-93c1-11e7-96cb-f0484386316a.png)

If I "materialize" the phantom types `I` and `O` in API, we get a different picture:

```ts
interface Api<I, O> {
    url: string;
    method: 'POST' | 'GET';
    __data: I;
    __dataO: O;
}

interface ApiDefinition {
    url: string;
}

interface GetApi<O> extends Api<{}, O> {
    method: 'GET';
}

type PostApi<I, O> = Api<I, O>;

function defineGetApi<O>(api: ApiDefinition) {
    return {
        url: api.url,
        method: 'GET'
    } as any as GetApi<O>;
}

function definePostApi<I, O>(api: ApiDefinition) {
    return {
        url: api.url,
        method: 'POST'
    } as any as PostApi<I, O>;
}

interface SampleInput {
    command: string;
}

interface SampleOutput {
    success: boolean;
}

const sampleGetApi = defineGetApi<SampleOutput>({
    url: 'string',
});
const samplePostApi = definePostApi<SampleInput, SampleOutput>({
    url: 'string',
});

function consumeApi<I, O>(api: Api<I, O>, input: I, onSuccess?: (output: O) => void) {
    return;
}

consumeApi(sampleGetApi, {}, (o) => o.success); // (1)
consumeApi(samplePostApi, {}, (o) => o.success); // (2)
```

Both compile. (1) should, inference works.
![image](https://user-images.githubusercontent.com/349397/30157045-83c337a2-93c1-11e7-84b9-5d371e4a61f6.png)

(Screenshots from TS 2.5)

**Comment by ahejlsberg** [comment]
This is working as intended and is an effect of TypeScript's structural type system. We can only make inferences for a type parameters if they actually occur somewhere in the _expansion_ of a generic type reference. See the FAQ:

https://github.com/Microsoft/TypeScript/wiki/FAQ#why-doesnt-type-inference-work-on-this-interface-interface-foot---

**Comment by alshain** [comment]
@ahejlsberg Thanks for the quick reply.

But why does it make a difference whether I `extend` the interface or not? 

![image](https://user-images.githubusercontent.com/349397/30187178-d32c9114-9428-11e7-957d-de1394a48337.png)

In the case of the post API, `O` doesn't appear in the expansion of the type either, unless I misunderstand something.



Here, `O` is inferred, from what I understand, through the phantom type.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.

**Comment by alshain** [comment]
From what I've gathered, the `extends` doesn't work because it's essentially syntactic sugar to define an interface. There is no actual connection between `A` and `B` in `A extends B` as there would be in a nominal type system, which is why the generic parameter isn't inferred.

Be that as it may,

> We can only make inferences for a type parameters if they actually occur somewhere in the expansion of a generic type reference. 

Doesn't that imply that what I did shouldn't work for a simple type alias either? It did infer the generic type merely based on the actual generic parameter, not on any member.
